### PR TITLE
IOS-13611 : Create a BOXRepresentationDownloadRequest.

### PR DESCRIPTION
**Pull request by crousselle**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @crousselle has signed the CLA. Thanks for the pull request!

**Comment by crousselle** [comment]
@boxtcarpel @thuyBox  FYI for the NSURLSession work, we're introducing a new BOXRequest.

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+File.h [hunk 1]
```diff
@@ -219,4 +221,31 @@
                                                                                 name:(NSString *)fileName
                                                                                 size:(NSUInteger)fileSize;
 
+/**
+ *  Generate a request to download a given representation of a file to a local filepath.
+ *
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+File.h:226]: Minor: Seems like you intended to align these, but didn't quite. Same on method below.

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//  BoxContentSDK
+//
+//  Created by Clement Rousselle on 3/2/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import "BOXRequestWithSharedLinkHeader.h"
+
+@class BOXRepresentation;
+
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h:12]: `NS_ASSUME_NONNULL`

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+- (instancetype)initWithLocalDestination:(NSString *)destinationPath
+                                  fileID:(NSString *)fileID
+                          representation:(BOXRepresentation *)representation;
+
+- (instancetype)initWithOutputStream:(NSOutputStream *)outputStream
+                              fileID:(NSString *)fileID
+                      representation:(BOXRepresentation *)representation;
+
+//Perform API request and any cache update only if refreshBlock is not nil
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h]: There's no `refreshBlock` here.

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+//
+//  Created by Clement Rousselle on 3/2/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import "BOXFileRepresentationDownloadRequest.h"
+#import "BOXRepresentation.h"
+#import "BOXAPIDataOperation.h"
+#import "BOXRequest_Private.h"
+
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:13]: `NS_ASSUME_NONNULL`

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                                                       successBlock:nil
+                                                       failureBlock:nil];
+    
+    BOXAssert(self.outputStream != nil || self.destinationPath != nil, @"An output stream or destination file path must be specified.");
+    BOXAssert(!(self.outputStream != nil && self.destinationPath != nil), @"You cannot specify both an outputStream and a destination file path.");
+    
+    if (self.outputStream != nil) {
+        dataOperation.outputStream = self.outputStream;
+    } else {
+        dataOperation.outputStream = [[NSOutputStream alloc] initToFileAtPath:self.destinationPath append:NO];
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:89]: With the move to `NSURLSession`, we'll probably want to specify the output file directly if we have one, to enable background downloads. Maybe that's not available on `BOXAPIDataOperation` on ToT, though.

File: BoxContentSDK/BoxContentSDKTests/BOXFileRepresentationDownloadRequestTests.m [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    [self expectationForNotification:BOXUserWasLoggedOutDueToErrorNotification object:nil handler:nil];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+}
+
+# pragma mark - private helpers
+
+- (NSData *)randomDataWithLength:(NSUInteger)length
+{
+    NSMutableData *mutableData = [NSMutableData dataWithCapacity: length];
+    for (unsigned int i = 0; i < length; i++) {
```
Comment (jlawton) [inline BoxContentSDK/BoxContentSDKTests/BOXFileRepresentationDownloadRequestTests.m:132]: Minor: Probably you can move by `sizeof(NSInteger)` and reduce to 25% or less of the runtime.

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+//
+//  BOXFileRepresentionListViewControllerTableViewController.h
+//  BoxContentSDKSampleApp
+//
+//  Created by Clement Rousselle on 3/2/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+
```
Comment (jlawton) [inline BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.h:10]: Nice! Sample app, too!

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
+{
+    return self.file.representations.count;
+}
+
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+    static NSString *UserTableViewCellIdentifier = @"representationCell";
+    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UserTableViewCellIdentifier];
```
Comment (jlawton) [inline BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m]: Could use new-style dequeue with class registration, and skip the `if`.

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            message = @"Could not download the representation";
+        } else {
+            title = @"Representation Downloaded !";
+            message = [NSString stringWithFormat:@"Asset Path : %@", finalPath];
+        }
+        
+        UIAlertController *controller = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
+        [controller addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (jlawton) [inline BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m]: Is there a reason to `dispatch_async` here?

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+File.h (hunk 1 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Clients/BOXContentClient+File.h:226]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h (hunk 2 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h:12]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h (hunk 3 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m (hunk 4 above)
Comment (crousselle) [inline BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m]: done

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m (hunk 5 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:89]: I don't think it is. I've flagged this change to tom and thuy to make sure required changes are made

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m (hunk 9 above)
Comment (crousselle) [reply BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m]: I assume the request callbacks are not made on the main thread. Maybe I'm wrong, I will double check

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m (hunk 8 above)
Comment (crousselle) [reply BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m]: yup good catch

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h (hunk 2 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.h:12]: actually no, we definitely can have some nil arguments here, especially blocks

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m (hunk 4 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:13]: same comment as before

File: BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m (hunk 9 above)
Comment (crousselle) [reply BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXFileRepresentionListViewControllerTableViewController.m]: ok we dispatch back on the main thread, so I assume we dispatch back on whatever thread the request was performed on. will remove the dispatch

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m (hunk 5 above)
Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:89]: Do we intend to support background download for this request? Would it be beneficial to have the content on disk upon app restarts? We would be able to add support for background download for the request with minimal change.

**Review by jdizoglio — APPROVED** [review]

File: BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m (hunk 5 above)
Comment (crousselle) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:89]: I think we need to support background downloads, yes. It should be handled similarly to our current download operation, I'm just unsure of what needs to be done for it.
Is it something we have to modify now, or that the URL session team will handle later on ? 
@thuyBox

Comment (thuyBox) [reply BoxContentSDK/BoxContentSDK/Requests/BOXFileRepresentationDownloadRequest.m:89]: it can be handled later, we'll keep record of it.

**Pull request merged**